File: script/check_csp.py
#!/usr/bin/env python3
"""
CSP チェックスクリプト (Python版)

目的:
 - ビルド成果物 (dist) に対して CSP に違反する可能性のあるコードが含まれていないかを検査します。
 - チェック項目: dist/index.html のインライン <script> / <style>、および dist 配下ファイルでの eval/new Function/Function の使用。

使い方:
 - 事前にビルドを行ってください: `pnpm build`（または `npm run build`）
 - スクリプトを実行: `python3 ./script/check_csp.py` または `./script/check_csp.py`（実行権限が必要）
"""

from __future__ import annotations

import os
import re
import sys
from typing import List, Tuple, Optional

INDEX_SCRIPT_RE = re.compile(r"<script[^>]*>[^<]", re.IGNORECASE)
INDEX_STYLE_RE = re.compile(r"<style\b[^>]*>\s*[^<\s]", re.IGNORECASE)
# Be case-sensitive: we want to detect expressions like "eval(" and "Function("
# but not the common lowercase "function(" tokens in minified bundles.
DANGEROUS_RE = re.compile(r"eval\(|new Function\(|\bFunction\(")


def find_inline_tags_in_index(
    index_path: str,
) -> Tuple[List[Tuple[int, str]], List[Tuple[int, str]]]:
    scripts: List[Tuple[int, str]] = []
    styles: List[Tuple[int, str]] = []

    try:
        with open(index_path, "r", encoding="utf-8", errors="replace") as fh:
            for lineno, line in enumerate(fh, start=1):
                if INDEX_SCRIPT_RE.search(line):
                    scripts.append((lineno, line.rstrip("\n")))
                if INDEX_STYLE_RE.search(line):
                    styles.append((lineno, line.rstrip("\n")))
    except FileNotFoundError:
        # index.html がない場合は空の一覧を返す（呼び出し側で存在チェック）
        return scripts, styles

    return scripts, styles


def find_dangerous_patterns(root_dir: str) -> List[Tuple[str, Optional[int], str]]:
    """dist 配下を再帰的にスキャンして危険なパターンを探す。

    戻り値: (相対パス, 行番号または None, 行の内容または説明)
    """
    matches: List[Tuple[str, Optional[int], str]] = []

    for dirpath, _, filenames in os.walk(root_dir):
        for fname in filenames:
            path = os.path.join(dirpath, fname)
            rel = os.path.relpath(path, root_dir)
            try:
                with open(path, "r", encoding="utf-8", errors="replace") as fh:
                    for lineno, line in enumerate(fh, start=1):
                        if DANGEROUS_RE.search(line):
                            matches.append((rel, lineno, line.rstrip("\n")))
            except Exception:
                # テキストとして読めない場合はバイナリとして簡易チェック
                try:
                    with open(path, "rb") as bf:
                        content = bf.read()
                        if (
                            b"eval(" in content
                            or b"new Function(" in content
                            or b"Function(" in content
                        ):
                            matches.append(
                                (rel, None, "<binary file contains pattern>")
                            )
                except Exception:
                    # 読めないファイルは無視
                    pass

    return matches


def main() -> None:
    print("Running CSP checks...")

    if not os.path.isdir("dist"):
        print(
            "Error: dist/ directory not found. Did you run the build?", file=sys.stderr
        )
        sys.exit(2)

    index_path = os.path.join("dist", "index.html")
    inline_scripts, inline_styles = find_inline_tags_in_index(index_path)
    eval_matches = find_dangerous_patterns("dist")

    if inline_scripts or inline_styles or eval_matches:
        print("CSP check failed:", file=sys.stderr)
        if inline_scripts:
            print("Inline <script> found in dist/index.html:", file=sys.stderr)
            for lineno, text in inline_scripts:
                print(f"{index_path}:{lineno}:{text}", file=sys.stderr)
        if inline_styles:
            print("Inline <style> found in dist/index.html:", file=sys.stderr)
            for lineno, text in inline_styles:
                print(f"{index_path}:{lineno}:{text}", file=sys.stderr)
        if eval_matches:
            print(
                "Potential dangerous patterns found in dist (eval/new Function/Function):",
                file=sys.stderr,
            )
            for rel, lineno, text in eval_matches:
                if lineno is None:
                    print(f"{rel}:<binary>: {text}", file=sys.stderr)
                else:
                    print(f"{rel}:{lineno}:{text}", file=sys.stderr)
        sys.exit(1)

    print("CSP checks passed.")
    sys.exit(0)


if __name__ == "__main__":
    main()

File: script/check_csp.ts
#!/usr/bin/env node

/**
 * CSP check script (TypeScript)
 *
 * Implements CSP checks for the built assets (originally implemented in Python).
 */

import fs from "fs";
import path from "path";

const INDEX_SCRIPT_RE = /<script(?![^>]*\bsrc\b)[^>]*>[^<]/i;
const INDEX_STYLE_RE = /<style\b[^>]*>\s*[^<\s]/i;
const DANGEROUS_RE = /eval\(|new Function\(|\bFunction\(/;

export function findInlineTagsInIndex(indexPath: string): [Array<[number, string]>, Array<[number, string]>] {
    const scripts: Array<[number, string]> = [];
    const styles: Array<[number, string]> = [];

    try {
        const text = fs.readFileSync(indexPath, { encoding: "utf-8" });
        const lines = text.split("\n");
        for (let i = 0; i < lines.length; i++) {
            const lineno = i + 1;
            const line = lines[i];
            if (INDEX_SCRIPT_RE.test(line)) {
                scripts.push([lineno, line.replace(/\n$/, "")]);
            }
            if (INDEX_STYLE_RE.test(line)) {
                styles.push([lineno, line.replace(/\n$/, "")]);
            }
        }
    } catch (err: unknown) {
        if ((err as NodeJS.ErrnoException).code === "ENOENT") {
            // Missing index is treated as empty lists (same as Python behaviour)
            return [scripts, styles];
        }
        throw err;
    }

    return [scripts, styles];
}

function hasPatternWithBoundary(content: Buffer, pattern: Buffer): boolean {
    let index = content.indexOf(pattern);
    while (index !== -1) {
        if (index === 0) return true;
        const prevByte = content[index - 1];
        if (prevByte < 128) {
            const prevChar = String.fromCharCode(prevByte);
            if (!(/[A-Za-z0-9_]/.test(prevChar))) {
                return true;
            }
        } else {
            // Non-ASCII byte treated as word boundary (conservative)
            return true;
        }
        index = content.indexOf(pattern, index + 1);
    }
    return false;
}

export function findDangerousPatterns(rootDir: string): Array<[string, number | null, string]> {
    const matches: Array<[string, number | null, string]> = [];

    function walk(dir: string) {
        const entries = fs.readdirSync(dir, { withFileTypes: true });
        for (const ent of entries) {
            const full = path.join(dir, ent.name);
            const rel = path.relative(rootDir, full);
            if (ent.isDirectory()) {
                walk(full);
                continue;
            }

            try {
                const text = fs.readFileSync(full, { encoding: "utf-8" });
                const lines = text.split("\n");
                for (let i = 0; i < lines.length; i++) {
                    const lineno = i + 1;
                    const line = lines[i];
                    if (DANGEROUS_RE.test(line)) {
                        matches.push([rel, lineno, line.replace(/\n$/, "")]);
                    }
                }
            } catch (e) {
                // Not readable as text, try binary read
                try {
                    const content = fs.readFileSync(full);
                    if (
                        content.includes(Buffer.from("eval(")) ||
                        content.includes(Buffer.from("new Function(")) ||
                        hasPatternWithBoundary(content, Buffer.from("Function("))
                    ) {
                        matches.push([rel, null, "<binary file contains pattern>"]);
                    }
                } catch {
                    // ignore unreadable files
                }
            }
        }
    }

    walk(rootDir);
    return matches;
}

export function main(): number {
    console.log("Running CSP checks...");

    if (!fs.existsSync("dist") || !fs.statSync("dist").isDirectory()) {
        console.error("Error: dist/ directory not found. Did you run the build?");
        return 2;
    }

    const indexPath = path.join("dist", "index.html");
    const [inlineScripts, inlineStyles] = findInlineTagsInIndex(indexPath);
    const evalMatches = findDangerousPatterns("dist");

    if (inlineScripts.length || inlineStyles.length || evalMatches.length) {
        console.error("CSP check failed:");
        if (inlineScripts.length) {
            console.error("Inline <script> found in dist/index.html:");
            for (const [lineno, text] of inlineScripts) {
                console.error(`${indexPath}:${lineno}:${text}`);
            }
        }
        if (inlineStyles.length) {
            console.error("Inline <style> found in dist/index.html:");
            for (const [lineno, text] of inlineStyles) {
                console.error(`${indexPath}:${lineno}:${text}`);
            }
        }
        if (evalMatches.length) {
            console.error(
                "Potential dangerous patterns found in dist (eval/new Function/Function):"
            );
            for (const [rel, lineno, text] of evalMatches) {
                if (lineno === null) {
                    console.error(`${rel}:<binary>: ${text}`);
                } else {
                    console.error(`${rel}:${lineno}:${text}`);
                }
            }
        }
        return 1;
    }

    console.log("CSP checks passed.");
    return 0;
}

if (!process.env.VITEST && process.env.NODE_ENV !== "test") {
    const code = main();
    process.exit(code);
}
